Source/WebCore/platform/graphics/ImageSource.cpp

```diff
@@ -341,95 +335,35 @@ bool ImageSource::canUseAsyncDecoding()
-                // The queue may have been closed if after we got the frame NativeImage, stopAsyncDecodingQueue() was called.
-                if (protectedDecodingQueue.ptr() == protectedThis->m_decodingQueue && protectedDecoder.ptr() == protectedThis->m_decoder) {
-                    ASSERT(protectedThis->m_frameCommitQueue.first() == frameRequest);
-                    protectedThis->m_frameCommitQueue.removeFirst();
-                    protectedThis->cachePlatformImageAtIndexAsync(WTFMove(platformImage), frameRequest.index, frameRequest.subsamplingLevel, frameRequest.decodingOptions);
-                } else
-                    LOG(Images, "ImageSource::%s - %p - url: %s [frame %ld will not cached]", __FUNCTION__, protectedThis.ptr(), sourceURL.utf8().data(), frameRequest.index);
-            });
-        }
-
-        // Ensure destruction happens on creation thread.
-        callOnMainThread([protectedThis = WTFMove(protectedThis), protectedQueue = WTFMove(protectedDecodingQueue), protectedDecoder = WTFMove(protectedDecoder)] () mutable { });
-    });
+    return *m_workQueue;
 }
 
-void ImageSource::requestFrameAsyncDecodingAtIndex(size_t index, SubsamplingLevel subsamplingLevel, const std::optional<IntSize>& sizeForDrawing)
+void ImageSource::requestFrameAsyncDecodingAtIndex(size_t index, SubsamplingLevel subsamplingLevel, ImageAnimatingState animatingState, const DecodingOptions& options)
 {
     ASSERT(isDecoderAvailable());
-    if (!hasAsyncDecodingQueue())
-        startAsyncDecodingQueue();
-
     ASSERT(index < m_frames.size());
-
-    LOG(Images, "ImageSource::%s - %p - url: %s [enqueuing frame %ld for decoding]", __FUNCTION__, this, sourceURL().string().utf8().data(), index);
-    DecodingOptions decodingOptions = { DecodingMode::Asynchronous, sizeForDrawing };
-    m_frameRequestQueue->enqueue({ index, subsamplingLevel, decodingOptions });
-    m_frameCommitQueue.append({ index, subsamplingLevel, decodingOptions });
-}
-
-bool ImageSource::isAsyncDecodingQueueIdle() const
-{
-    return m_frameCommitQueue.isEmpty();
+    workQueue().dispatch({ static_cast<unsigned>(index), subsamplingLevel, animatingState, options });
 }
 
-void ImageSource::stopAsyncDecodingQueue()
+void ImageSource::stopWorkQueue()
 {
-    if (!hasAsyncDecodingQueue())
+    if (!m_workQueue)
         return;
 
-    std::for_each(m_frameCommitQueue.begin(), m_frameCommitQueue.end(), [this](const ImageFrameRequest& frameRequest) {
-        ImageFrame& frame = m_frames[frameRequest.index];
-        if (!frame.isInvalid()) {
-            LOG(Images, "ImageSource::%s - %p - url: %s [decoding has been cancelled for frame %ld]", __FUNCTION__, this, sourceURL().string().utf8().data(), frameRequest.index);
-            frame.clear();
-        }
-    });
+    LOG(Images, "ImageSource::%s - %p - url: %s. Decoding work queue will be stopped.", __FUNCTION__, this, sourceUTF8());
+
+    m_workQueue->stop();
```

Comment:
You need a Ref/RefPtr on the stack before you call `stop()`, e.g. `RefPtr { m_workQueue }->stop()` or `protectedWorkQueue()->stop()`